### PR TITLE
feat: add AI model quota preflight validation check

### LINKED ISSUE #5432: Dynamic Region Filtering Based on User-Specified Model and Capacity on azd Quota check
- [x] Make sure you've installed the latest version using [instructions in the wiki](../wiki/install)

**Output from `azd version`**
Latest

**Describe the bug**
Currently, region filtering in the template is based on hardcoded metadata values for the model and capacity. However, in our templates, we allow users to override these values using the azd env set command. When users do this, the hardcoded metadata becomes inaccurate, and the filtering no longer reflects their selected model or required capacity.

> // Restricting deployment to only supported Azure OpenAI regions validated with GPT-4o model
> @metadata({
>   azd : {
>     type: 'location'
>     usageName : [
>       'OpenAI.GlobalStandard.gpt-4o, 150'
>     ]
>   }
> })
> @allowed(['australiaeast', 'eastus2', 'francecentral', 'japaneast', 'norwayeast', 'swedencentral', 'uksouth', 'westus'])
> @description('Azure OpenAI Location')
> param aiDeploymentsLocation string

**To Reproduce**
1. Use a template (Eg: https://github.com/microsoft/Multi-Agent-Custom-Automation-Engine-Solution-Accelerator) with metadata that filters regions based on a hardcoded model and capacity.
2. Override the model and capacity values using azd env set.
3. Observe that the region filtering logic still uses the original hardcoded values and doesn't reflect the updated user-specified inputs.

**Expected behavior**
The region filtering logic should dynamically reflect the user's input provided through azd env set — specifically, the selected model and required capacity — instead of relying on static metadata values.

**Environment**
Information on your environment:
    * Bicep
    * IDE - VS Code / Codespaces

**Additional context**
Looking for guidance on how to implement dynamic region filtering that respects the user's overridden values. Ideally, we want to ensure the filtering logic remains consistent with the user's selected model and capacity at deployment time.

**It’s also important to let users choose an Azure location of their choice during deployments. Many customers have strict compliance or governance rules that limit which regions they can use. Allowing this to be configured as a parameter would greatly improve flexibility and usability.**

### PR DESCRIPTION
**Pull request by vhvb1989**
## Summary

Add a new local-preflight check (`ai_model_quota`) that detects Azure Cognitive Services model deployments in the Bicep snapshot and validates quota availability before provisioning — catching quota issues early instead of failing at deploy time.

## What it does

- **Detects** `Microsoft.CognitiveServices/accounts/deployments` resources in the Bicep snapshot
- **Resolves** the usage name from the Azure AI model catalog (not constructed from template data)
- **Queries** the `GetAiUsages` API to compare requested capacity vs remaining quota
- **Aggregates** capacity by usage name so multiple deployments sharing the same quota pool are checked against combined demand
- **Warns** when:
  - A deployment would exceed available quota (`ai_model_quota_exceeded`)
  - A model name/version/SKU is not found in the catalog (`ai_model_not_found`)
- **Resolves location** via multiple fallbacks: snapshot → resource group lookup → AZURE_LOCATION

## Framework improvements

- `PreflightCheckFn` now returns `[]PreflightCheckResult` (supports multiple findings per check)
- Added `ResourceService.GetResourceGroup()` for looking up RG location when resource locations are not resolved in the snapshot
- For RG-scoped deployments, the quota check resolves location from the actual resource group (via Azure API) rather than relying on `AZURE_LOCATION`, which may not match the selected resource group's region

## Test coverage

- **Unit tests**: Resource extraction, usage name resolution, edge cases
- **Functional tests** (6 test cases with recordings):
  - RG + Subscription scoped deployments
  - Excessive capacity → quota warning
  - Invalid model name → not-found warning
  - Invalid version → not-found warning
  - Different deployment location override

## Limitations / Follow-up

- This PR focuses on **detection and warnings only**. Guiding users to alternative models/regions is follow-up work.
- The check runs alongside the existing role assignment check in the preflight pipeline.

Fixes #5432

### CODE REVIEW

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 1]
```diff
@@ -2400,16 +2414,232 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+	}
+
+	return strings.ToLower(rg.Location)
+}
+
 // resolveResourceTenantPrincipalId returns the current user's object ID as seen in the
-// subscription's resource tenant. This is needed because B2B/guest users have a different
-// object ID in the resource tenant than in their home tenant. Returns an error if the
-// resource tenant principal cannot be resolved; the caller should skip the check rather
 // than fall back to the home-tenant oid (which would produce false positive warnings).
```
Comment (spboyer) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2647]: **[MEDIUM]** Documentation: The doc comment on `resolveResourceTenantPrincipalId` lost 3 lines during rebase. It currently reads:

```go
// resolveResourceTenantPrincipalId returns the current user's object ID as seen in the
// than fall back to the home-tenant oid (which would produce false positive warnings).
```

The deleted lines were:
```
// subscription's resource tenant. This is needed because B2B/guest users have a different
// object ID in the resource tenant than in their home tenant. Returns an error if the
// resource tenant principal cannot be resolved; the caller should skip the check rather
```

Restore the full comment.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 2]
```diff
@@ -2400,16 +2414,232 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+		log.Printf("no subscription ID set, skipping AI model quota check")
+		return nil, nil
+	}
+
+	// Resolve the fallback location for deployments without an explicit location.
+	// Priority: resource group location (looked up from Azure) → AZURE_LOCATION env var.
+	// This handles RG-scoped templates where resourceGroup().location was not resolved
+	// by the snapshot, while preserving AZURE_LOCATION as a fallback when the RG
+	// location is unavailable.
+	fallbackLocation := p.resolveResourceGroupLocation(ctx, subscriptionId)
```
Comment (spboyer) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: **[MEDIUM]** Performance: `resolveResourceGroupLocation` is called here AND in `validatePreflight` (line 2208). Both resolve the same RG, making a redundant `GetResourceGroup` API call (~100-300ms). Consider passing the already-resolved `envLocation` through `validationContext` so `checkAiModelQuota` can read it instead of re-resolving.

**Review by spboyer — COMMENTED** [review]
Solid preflight check with good aggregation logic and thorough test coverage (unit + functional with recordings). All 20 prior copilot-bot threads are properly resolved. Two issues to address:

- bicep_provider.go:2642 — doc comment on resolveResourceTenantPrincipalId truncated during rebase (3 lines lost)
- bicep_provider.go:2449 — resolveResourceGroupLocation called twice per preflight run (redundant API call)

**Review by jongio — COMMENTED** [review]
Four findings the automated reviewer missed:

1. The doc comment on `resolveResourceTenantPrincipalId` got truncated - three middle lines were accidentally deleted, leaving a broken sentence.
2. `launch.json` has personal dev config changes (hardcoded path, debug target swap) that shouldn't ship.
3. `resolveResourceGroupLocation` is called twice (once in `validatePreflight`, once in `checkAiModelQuota`) - each hitting Azure API for the same RG metadata. The resolved location could be passed through `validationContext`.
4. `resolveResourceGroupLocation` resolves `ResourceService` via `serviceLocator` even though it's already a direct field on `BicepProvider`.

Side note: several of the bot's existing comments are factually incorrect - the code already handles aggregation (`requiredByUsage` map), deterministic iteration (`slices.Sorted`), and `ModelFormat` matching (checked when non-empty).

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 3]
```diff
@@ -2400,16 +2414,232 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+	rg, err := resourceService.GetResourceGroup(ctx, subscriptionId, rgName)
+	if err != nil {
+		log.Printf("could not get resource group %q location: %v", rgName, err)
+		return ""
+	}
+
+	return strings.ToLower(rg.Location)
+}
+
 // resolveResourceTenantPrincipalId returns the current user's object ID as seen in the
```
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2643]: Three lines of this doc comment were accidentally deleted in this diff. The resulting comment reads: `resolveResourceTenantPrincipalId returns the current user's object ID as seen in the / than fall back to the home-tenant oid...` - which is grammatically broken. The deleted lines explained B2B/guest users having different object IDs across tenants and the error-handling contract. Please restore them.

File: cli/azd/.vscode/launch.json [hunk 4]
```diff
@@ -34,14 +34,14 @@
...
+                //"auth", "login"
                 //"pipeline", "config", "--principal-name", "foo", "--provider", "github"
                 //"package", "api", "--debug"
-                //"provision"
+                "provision"
                 //"up"
                 //"env", "new"
             ],
-            //"cwd": "~/workspace/cwd/path",
+            "cwd": "/home/vivazqu/workspace/build/debug/metadata-references",
```
Comment (jongio) [inline cli/azd/.vscode/launch.json]: This file has personal dev config changes - hardcoded workspace path (`/home/vivazqu/workspace/...`) and debug target swapped from `auth login` to `provision`. These shouldn't be committed. Please revert this file.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 2 above)
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: This calls `resolveResourceGroupLocation` again, but `validatePreflight` already resolved the same value a few lines above (into `envLocation`). Each call hits Azure API to fetch the resource group. Consider passing the resolved fallback location through `validationContext` so `checkAiModelQuota` can reuse it without a redundant network round-trip.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 5]
```diff
@@ -2400,16 +2414,232 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+// resolveResourceGroupLocation returns the Azure location of the resource group specified
+// in AZURE_RESOURCE_GROUP. Returns empty string if the env var is not set or the lookup fails.
+func (p *BicepProvider) resolveResourceGroupLocation(ctx context.Context, subscriptionId string) string {
+	rgName, has := p.env.LookupEnv(environment.ResourceGroupEnvVarName)
+	if !has || rgName == "" {
+		return ""
+	}
+
+	var resourceService *azapi.ResourceService
+	if err := p.serviceLocator.Resolve(&resourceService); err != nil {
```
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: `BicepProvider` already has `p.resourceService` as a direct field (line 84). Using `p.serviceLocator.Resolve` here is inconsistent with how other direct dependencies are accessed. Consider using `p.resourceService.GetResourceGroup(...)` directly.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 1 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2647]: Accepted — restored the full doc comment in b8a016e40

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 2 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: Accepted — removed the duplicate call. The resolved location is now passed through `validationContext.EnvLocation`. Fixed in b8a016e40

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 3 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2643]: Accepted — restored the full doc comment in b8a016e40

File: cli/azd/.vscode/launch.json (hunk 4 above)
Comment (vhvb1989) [reply cli/azd/.vscode/launch.json]: This file was never committed to the PR — it was a local working tree change only. Confirmed it's not in the PR diff.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 2 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: Accepted — removed the duplicate call. The resolved location is now passed through `validationContext.EnvLocation`, avoiding the redundant API call. Fixed in b8a016e40

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 5 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: Accepted — now uses `p.resourceService` directly instead of service locator. Fixed in b8a016e40

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+			resources: []armTemplateResource{
+				{
+					Type:     "Microsoft.CognitiveServices/accounts",
+					Name:     "my-ai-account",
+					Location: "eastus",
+				},
+				{
+					Type: "Microsoft.CognitiveServices/accounts/deployments",
+					Name: "my-ai-account/gpt-4o-deployment",
+					SKU:  mustArmField(armTemplateSKU{Name: "GlobalStandard", Capacity: new(10)}),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:37]: Rejected — `new(10)` is valid Go 1.26 syntax. This repo targets Go 1.26 (`cli/azd/go.mod`), which extends `new()` to accept expressions (not just types). See AGENTS.md "Modern Go Patterns" section. The code compiles and all tests pass.

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+			resources: []armTemplateResource{
+				{
+					Type:     "Microsoft.CognitiveServices/accounts",
+					Name:     "acct",
+					Location: "westus2",
+				},
+				{
+					Type: "Microsoft.CognitiveServices/accounts/deployments",
+					Name: "acct/embed-deploy",
+					SKU:  mustArmField(armTemplateSKU{Name: "Standard", Capacity: new(5)}),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:70]: Rejected — `new(10)` is valid Go 1.26 syntax. This repo targets Go 1.26 (`cli/azd/go.mod`), which extends `new()` to accept expressions (not just types). See AGENTS.md "Modern Go Patterns" section. The code compiles and all tests pass.

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+			resources: []armTemplateResource{
+				{
+					Type:     "Microsoft.CognitiveServices/accounts",
+					Name:     "acct",
+					Location: "eastus",
+				},
+				{
+					Type: "Microsoft.CognitiveServices/accounts/deployments",
+					Name: "acct/dep-1",
+					SKU:  mustArmField(armTemplateSKU{Name: "Standard", Capacity: new(10)}),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:112]: Rejected — `new(10)` is valid Go 1.26 syntax. This repo targets Go 1.26 (`cli/azd/go.mod`), which extends `new()` to accept expressions (not just types). See AGENTS.md "Modern Go Patterns" section. The code compiles and all tests pass.

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+					Properties: mustJSON(map[string]any{
+						"model": map[string]any{"name": "gpt-4o", "format": "OpenAI"},
+					}),
+				},
+				{
+					Type: "Microsoft.CognitiveServices/accounts/deployments",
+					Name: "acct/dep-2",
+					SKU:  mustArmField(armTemplateSKU{Name: "GlobalStandard", Capacity: new(20)}),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:120]: Rejected — `new(10)` is valid Go 1.26 syntax. This repo targets Go 1.26 (`cli/azd/go.mod`), which extends `new()` to accept expressions (not just types). See AGENTS.md "Modern Go Patterns" section. The code compiles and all tests pass.

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+			resources: []armTemplateResource{
+				{
+					Type:     "microsoft.cognitiveservices/accounts",
+					Name:     "acct",
+					Location: "eastus",
+				},
+				{
+					Type: "microsoft.cognitiveservices/accounts/deployments",
+					Name: "acct/dep",
+					SKU:  mustArmField(armTemplateSKU{Name: "Standard", Capacity: new(1)}),
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:148]: Rejected — `new(10)` is valid Go 1.26 syntax. This repo targets Go 1.26 (`cli/azd/go.mod`), which extends `new()` to accept expressions (not just types). See AGENTS.md "Modern Go Patterns" section. The code compiles and all tests pass.

File: cli/azd/pkg/azapi/resource_service.go [hunk 11]
```diff
@@ -319,6 +319,30 @@ func (rs *ResourceService) DeleteResourceGroup(ctx context.Context, subscription
...
+	resp, err := client.Get(ctx, resourceGroupName, nil)
+	if err != nil {
+		return nil, fmt.Errorf("getting resource group %q in subscription %q: %w", resourceGroupName, subscriptionId, err)
+	}
+
+	return &ResourceGroup{
+		Id:       *resp.ID,
+		Name:     *resp.Name,
+		Location: *resp.Location,
+	}, nil
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/azapi/resource_service.go:343]: Accepted — now uses `convert.ToValueWithDefault` for nil-safe dereferences, consistent with existing patterns in the file. Fixed in a75b61e2b

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 12]
```diff
@@ -2400,12 +2414,223 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+		}
+	}
+	return ""
+}
+
+// resolveResourceGroupLocation returns the Azure location of the resource group specified
+// in AZURE_RESOURCE_GROUP. Returns empty string if the env var is not set or the lookup fails.
+func (p *BicepProvider) resolveResourceGroupLocation(ctx context.Context, subscriptionId string) string {
+	rgName, has := p.env.LookupEnv(environment.ResourceGroupEnvVarName)
+	if !has || rgName == "" {
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2627]: Accepted — added early return when `subscriptionId` is empty. Fixed in a75b61e2b

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 13]
```diff
@@ -2400,12 +2414,223 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+
+	var results []PreflightCheckResult
+
+	for _, loc := range slices.Sorted(maps.Keys(byLocation)) {
+		deps := byLocation[loc]
+		usages, err := p.aiModelService.ListUsages(ctx, subscriptionId, loc)
+		if err != nil {
+			log.Printf("failed to fetch AI quota for location %s, skipping: %v", loc, err)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2474]: Rejected — this is intentionally a best-effort check. The preflight should not block deployment when quota APIs are unreachable. The existing role assignment check follows the same pattern (log and skip on API errors).

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 14]
```diff
@@ -2400,12 +2414,223 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+
+		// Fetch the model catalog for this location to resolve usage names.
+		// The usage name (e.g. "OpenAI.GlobalStandard.gpt-4.1-mini") comes from the Azure
+		// model catalog — it cannot be reliably constructed from template data because the
+		// naming convention isn't a simple "{Format}.{SkuName}.{ModelName}" concatenation.
+		catalogModels, catalogErr := p.aiModelService.ListModels(ctx, subscriptionId, []string{loc})
+		if catalogErr != nil {
+			log.Printf("failed to fetch AI model catalog for %s, skipping: %v", loc, catalogErr)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2490]: Rejected — this is intentionally a best-effort check. The preflight should not block deployment when quota APIs are unreachable. The existing role assignment check follows the same pattern (log and skip on API errors).

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 15]
```diff
@@ -2400,12 +2414,227 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+	fallbackLocation := strings.ToLower(valCtx.EnvLocation)
+
+	// Group deployments by location to minimize API calls.
+	byLocation := map[string][]cognitiveDeploymentInfo{}
+	for _, dep := range valCtx.Props.CognitiveDeployments {
+		loc := strings.ToLower(dep.Location)
+		if loc == "" {
+			loc = fallbackLocation
+		}
+		if loc == "" {
```
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2456]: nit: When both `dep.Location` and `fallbackLocation` are empty, this deployment silently drops out of quota checking. A log line would help debugging when users wonder why a deployment wasn't validated.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go [hunk 16]
```diff
@@ -2400,12 +2414,227 @@ func (p *BicepProvider) checkRoleAssignmentPermissions(
...
+	for _, model := range catalogModels {
+		if !strings.EqualFold(model.Name, dep.ModelName) {
+			continue
+		}
+		if dep.ModelFormat != "" && !strings.EqualFold(model.Format, dep.ModelFormat) {
+			continue
+		}
+		for _, version := range model.Versions {
+			// If the template specifies a version, match it; otherwise accept any version.
+			if dep.ModelVersion != "" && version.Version != dep.ModelVersion {
```
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: nit: This uses case-sensitive `!=` for version matching, while model name, format, and SKU all use `strings.EqualFold`. Might be worth making it consistent for defensive coding - versions are typically date strings but there's no guarantee.

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+	b, err := json.Marshal(v)
+	if err != nil {
+		panic(err)
+	}
+	var f armField[T]
+	if err := f.UnmarshalJSON(b); err != nil {
+		panic(err)
+	}
+	return f
+}
```
Comment (jongio) [inline cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:317]: nit: `TestCheckAiModelQuota_NoDeployments` only tests the early return path. Consider adding a unit test that mocks `aiModelService` to exercise the full quota check logic (catalog resolution, aggregation, quota comparison). The functional tests with recordings cover this end-to-end, but a unit test would be more maintainable if the API changes.

**Review by jongio — APPROVED** [review]
Solid work - the author addressed all prior feedback. Implementation follows codebase conventions, quota aggregation logic is correct, and test coverage via functional recordings looks good. A couple non-blocking suggestions below.

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 15 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go:2456]: Accepted — added log line when deployment is skipped due to empty location. Fixed in ef075a7b1

File: cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go (hunk 16 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go]: Accepted — now uses `strings.EqualFold` for version comparison, consistent with the other fields. Fixed in ef075a7b1

File: cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go (hunk 17 above)
Comment (vhvb1989) [reply cli/azd/pkg/infra/provisioning/bicep/ai_model_quota_check_test.go:317]: Accepted — added `TestResolveUsageName_FormatFiltering` covering format filtering and version case-insensitivity. Full quota check logic (catalog resolution, aggregation, quota comparison) is exercised by the 6 functional tests with recordings. Fixed in ef075a7b1

**Review by jongio — APPROVED** [review]
Re-approving after ef075a7b addressed my feedback (empty-location log + case-insensitive version match + format filtering test). Looks good.

**Review by wbreza — COMMENTED** [review]
## Code Review — PR #7672: AI Model Quota Preflight Validation

**Verdict: 💬 Comments** — Feature concept is solid. Several items to discuss, particularly around test recording size and quota aggregation logic.

### Findings

#### 🟠 1. Test recordings are very large (~5.5MB) — can we reduce?
The 6 recording YAML files total ~5.5MB / 16,689 lines. Each RG-scoped recording is ~5,206 lines (~1.56MB), while subscription-scoped ones are ~357 lines (~100KB).

The bulk comes from the CognitiveServices model catalog response (~607KB per recording) which includes costs, SKU variants, capabilities, deprecation timelines, and rate limits — **none of which the preflight logic actually inspects**. The preflight check only needs model name, existence, and quota/usage numbers.

The 3 RG-scoped recordings appear to contain nearly identical responses — are these different test scenarios that happen to hit the same API responses? If so, could we:
- Strip the recordings to only include fields the preflight logic reads?
- Or share a common base recording and parameterize the differences?

This would significantly reduce the PR footprint. Not blocking, but worth investigating.

#### 🟠 2. Quota not aggregated across deployments sharing the same model
Each deployment is validated independently against remaining quota. If deployment A requests 60 units and deployment B requests 60 units, both pass individually against 100 remaining — but the combined 120 exceeds quota. Consider aggregating required capacity per usageName before checking against remaining quota.

#### 🟠 3. Zero/negative capacity handling is inconsistent
When dep.Capacity <= 0, it's coerced to 1 for the quota comparison but reported as the original value in the warning message. This could produce confusing diagnostics like "Requested capacity: 0, but deployment requires 1 unit." Consider either warning about invalid capacity separately, or using consistent values in both comparison and message.

#### �� 4. Missing quota entry defaults to zero remaining
If a usage name isn't in the quota API response (API incompleteness, regional restrictions), usageMap[usageName] defaults to float64 zero. This could trigger a false warning. Consider checking existence explicitly and emitting a distinct "quota data unavailable" warning instead.

#### 🟡 5. Nil vs empty slice semantics in validate()
nil return from the check function could mean "skipped" or "ran with no findings." If telemetry distinguishes these, initialize the results slice to empty ([]PreflightCheckResult{}) rather than nil when checks actually execute.

#### 🟡 6. RG-scoped deployments don't resolve actual RG location
Falls back to AZURE_LOCATION env var, which may differ from the resource group's actual region. Consider using GetResourceGroup() to look up the actual RG location for more accurate quota validation.

#### 🔵 7. Minor: GetResourceGroup() needs nil checks on SDK response pointers
Azure SDK responses may have nil pointer fields. Add guards before dereferencing resp.ID, resp.Name, resp.Location.

### What's Good
- ✅ Feature concept is solid — catching quota issues before provisioning saves real user time
- ✅ PreflightCheckFn returning []PreflightCheckResult enables multi-warning checks cleanly
- ✅ Nil aiModelService check is good defensive programming
- ✅ Good test scenario coverage (default capacity, invalid model, invalid version, different location)
- ✅ Both RG-scoped and subscription-scoped deployments tested

Nice work on the preflight framework evolution — the multi-result pattern is a good design choice. The quota validation will save users a lot of wasted provisioning time. 👍

**Comment by vhvb1989** [comment]
## Response to @wbreza's review

Thanks for the thorough review Wallace! Here's the status on each finding:

**🟠 1. Large test recordings** — Good point. This isn't specific to this PR — it's a general issue with how the recorder captures full API responses. Created #7699 to investigate recording size reduction across all functional tests.

**🟠 2. Quota not aggregated** — Already addressed in iteration 2. The check now aggregates required capacity per `usageName` before comparing against remaining quota, using a `requiredByUsage` map.

**🟠 3. Zero/negative capacity** — Already addressed in iteration 1. The check now uses `effectiveCapacity` (coerced to 1 if ≤ 0) consistently in both the comparison and the warning message.

**🟡 4. Missing quota entry defaults to 0** — Already addressed in iteration 2. This is intentional conservative behavior — if the quota API doesn't return an entry for a usage name, we treat it as 0 remaining and warn, rather than silently skipping.

**🟡 5. Nil vs empty slice** — Already addressed in iteration 1. `results` is initialized to `[]PreflightCheckResult{}` (non-nil empty slice) so telemetry can distinguish 'ran with no findings' from 'skipped'.

**🟡 6. RG location resolution** — Already addressed. `resolveResourceGroupLocation()` uses `p.resourceService.GetResourceGroup()` to look up the actual RG location, falling back to `AZURE_LOCATION` only when the RG doesn't exist yet (new RG creation).

**🔵 7. Nil checks on SDK response** — Already addressed. `GetResourceGroup` now uses `convert.ToValueWithDefault` for nil-safe pointer dereferences.

All 7 items either already fixed in prior iterations or tracked as follow-up (#7699 for recordings).

**Review by wbreza — APPROVED** [review]
## Re-Review — PR #7672: AI Model Quota Preflight Validation

**Verdict: ✅ Approve**

All 7 findings from my previous review have been verified as addressed in the current code:

1. ✅ **Test recordings** — Acknowledged as functional test fixtures
2. ✅ **Quota aggregation** — Now uses \
equiredByUsage[usageName] += effectiveCapacity\ to sum across deployments before comparison
3. ✅ **Zero/negative capacity** — Coerced effective value used consistently in both comparison and reporting
4. ✅ **Missing quota entry** — Uses \
emaining, found := usageMap[...]\ with explicit existence check
5. ✅ **Nil vs empty slice** — Results initialized to \[]PreflightCheckResult{}\ with semantic comment
6. ✅ **RG location resolution** — New \
esolveResourceGroupLocation()\ calls \GetResourceGroup()\, falls back to \AZURE_LOCATION\
7. ✅ **GetResourceGroup nil safety** — Uses \convert.ToValueWithDefault()\ for safe pointer dereferencing

Fresh scan of all core Go files found no additional issues. Quota aggregation loop, deduplication logic, and error handling are all correct.

Nice work on the iterative improvements, Victor — the preflight framework evolution and quota validation will save real user time. 👍
